### ISSUE TITLE
[bug] Conan should unset VSINSTALLDIR before calling vcvars

### ISSUE THREAD

**Issue by fredizzimo** [conan-io/conan#13977]
### Environment details

Windows + Visual Studio

### Steps to reproduce

This occurs in many different places, but here's one example. I'm experiencing it when calling `conan install` (actually through `BinaryInstaller` using a conan profile that uses a different version of Visual Studio than the one you are currently in. 

When a package is installed that uses CMakeDeps and the Ninja generator, a `conanvcvars.bat` file is generated, which is invoked when actually building. That file calls the `vcvarsall.bat` file corresponding to the Visual Studio version defined in the profile. 

This only works if `VSINSTALLDIR` is not set. `vcvarsall.bat` internally calls `vsdevcmd_start.bat` which has the following block of code.
```bat
if "%VSINSTALLDIR%" NEQ "" goto :set_devenvdir
@REM Set VSINSTALLDIR to Visual Studio Installation Directory
@REM vsdevcmd_start.bat location: Microsoft Visual Studio 16.0\Common7\Tools\vsdevcmd\core
@REM VSINSTALLDIR location: Microsoft Visual Studio 16.0
pushd "%~dp0..\..\..\..\"
set "VSINSTALLDIR=%CD%\"
popd
```

So it uses `VSINSTALLDIR` if it's set, and otherwise automatically detects it from the location of the `bat` file. `VSINSTALLDIR` is then used for finding extensions like location of compilers, cmake and so on. Since the list of extensions is determined by which `vcvarsall.bat` file you call, the resulting effect is that you get a frankenstein combination of Visual Studio, and unless you have the exact same extensions installed in both versions there will be errors.

The solution is simple, just undef the `VSINSTALLDIR` environment variable before calling `vcvarsall.bat`. At least for the purpose of CMake this is enough to make it work. But I haven't read the bat files in detail, there could be some other environment variables worth undefining too.

**Comment by memsharded** [comment]
Hi @fredizzimo 

Thanks for your report.
I thought that activating a vcvars when you are already in a VS-activated prompt was not a good idea, and as such, the recommendation was run in regular cmd terminals, and the packages activate the environment as needed. So I am a a bit concerned that the unset VSINSTALLDIR won't have other undesired unexpected side effects. I have had a look, but it is not evident in the docs, do you have some links that illustrate or document the usage and effects in this variable, and if activating a ``vcvars`` environment once you are already in an activated environment is possible or not?

**Comment by fredizzimo** [comment]
No, unfortunately I don't have any documentation, all of this is based on reading the bat files provided by Visual Studio and experimentation. 

As long as we don't let anything leak anything out the the calling process it should be fine, 
Variables like set `DevEnvDir`, `VisualStudioVersion` and so on are always overwritten, Variables like `PATH`, `LIB` and so on are prepended to, so the environment in the `vcvarsall.bat` should take precedence and it works.

The alternative is to block calling `vcvarsall`, and use the wrong compiler for installing the dependencies, but I don't like that idea. 

I don't know why `vcvarsall` tries to use an existing `VSINSTALLDIR` if not set. It must be used for some very special setups, maybe Microsoft internal testing. But given that it's not set globally on a system, I think it's quite safe to assume that `vcvarsall` should work without that variable.

I just noticed that `DEVENVDIR` is also re-used in a similar way, so that should be unset as well. But nothing else from what I can see. 

**Edit:** Note it's not critical for us to get this fixed, we have a local workaround, where we undefine the variables before calling `install` and that works for us at least.

**Comment by memsharded** [comment]
It seems that Microsoft discourages this flow:

https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170#developer_command_file_locations

> If you prefer to set the build environment in an existing command prompt window, you can use one of the command files created by the installer. We recommend you set the environment in a new command prompt window. **We don't recommend you later switch environments in the same command window.**

**Comment by fredizzimo** [comment]
In that case I suggest that you detect if it's already running inside Visual Studio or a developer command prompt, since figuring out why it was failing was not trivial. 

We will continue using our workaround of undefining whatever is needed, since requiring everyone to run `conan install` from a command prompt would be very inconvenient, compared to just open the folder in Visual Studio and hit build. It's also error-prone, since you have to remember to do it every time you fetch a new version from git, otherwise the dependencies might not update properly.

**Comment by memsharded** [comment]
> In that case I suggest that you detect if it's already running inside Visual Studio or a developer command prompt, since figuring out why it was failing was not trivial.

Yes, probably it is worth to print at least some warnings

> compared to just open the folder in Visual Studio and hit build. 

Could you please clarify what is your workflow? I thought you were using CMake + Ninja generator, and the flow was probably a bit using the command line, including the activation of the environment, which would include calling the ``conanvcvars.bat``. If you are not doing it, then the ``conanvcvars.bat`` is not being called?

**Comment by fredizzimo** [comment]
> Could you please clarify what is your workflow? 

We have our own "workspaces" implementation based on CMake. As a part of the cmake configure step we run our own Python based tool that prepares all the packages in the workspace to be built in editable mode and installs all the dependencies. It also generates an cmake include file that contains all the projects and adds the actual projects using the cmake `add_subdirectory` command .

We furthermore generate a global `CMakePresets.json`, which defines the Visual Studio toolchain to use, and different build configurations for example. So everyone has the same consistent configuration.

This works very well for us, and if you wish to you can use just the Visual Studio open folder to build and run without manually running any command-line commands. We also have another mode of compilation that is command-line based and basically just runs `conan install`, but that's an alternative workflow, and mostly used to verify that the packages works as regular non-editable packages as well.

`conanvcvars.bat` get's called when installing external dependencies, but since it's run by cmake started by Visual Studio some of the environment variables are already defined and it won't work. Our current workaround is to just undefine the variables inside our Python code, so that it behaves as if it was invoked in a clean environment.
